src/video/renderers/opengl3/gl3_renderer.c

```diff
@@ -0,0 +1,381 @@
+        vga_state_mark_remaps_flushed();
+    }
+}
+
+static inline void finish_offscreen(gl3_context *ctx) {
+    object_array_finish(ctx->objects);
+
+    // Set to VGA emulation state, and render to an indexed surface
+    glViewport(0, 0, NATIVE_W, NATIVE_H);
+    object_array_batch batch;
+    object_array_begin(ctx->objects, &batch);
+    activate_program(ctx->palette_prog_id);
+    render_target_activate(ctx->target);
+
+    object_array_blend_mode mode;
+    while(object_array_get_batch(ctx->objects, &batch, &mode)) {
+        video_set_blend_mode(ctx, mode);
+        object_array_draw(ctx->objects, &batch);
+    }
+}
+
+/**
+ * Disable render target, and dump its contents as RGBA to the screen.
+ */
+static inline void finish_onscreen(gl3_context *ctx) {
+    render_target_deactivate();
+    set_screen_viewport(ctx);
+    video_set_blend_mode(ctx, MODE_SET);
+    activate_program(ctx->rgba_prog_id);
+    if(ctx->draw_atlas) {
+        bind_uniform_1i(ctx->rgba_prog_id, "framebuffer", TEX_UNIT_ATLAS);
+    } else {
+        bind_uniform_1i(ctx->rgba_prog_id, "framebuffer", TEX_UNIT_FBO);
+    }
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
+}
+
+static void render_finish(void *userdata) {
+    gl3_context *ctx = userdata;
+    flush_palettes(ctx);
+    flush_remaps(ctx);
+    finish_offscreen(ctx);
+    finish_onscreen(ctx);
+
+    // Snap screenshot from the freshly rendered state.
+    if(ctx->screenshot_cb) {
+        capture_screenshot(ctx);
+        ctx->screenshot_cb = NULL;
+    }
+
+    // Flip buffers. If vsync is off, we should sleep here
+    // so hat our main loop doesn't eat up all cpu :)
+    SDL_GL_SwapWindow(ctx->window);
+}
+
+static void render_area_prepare(void *userdata, const SDL_Rect *area) {
+    gl3_context *ctx = userdata;
+    object_array_prepare(ctx->objects);
+    ctx->culling_area = *area;
+}
+
+static void render_area_finish(void *userdata, surface *dst) {
+    gl3_context *ctx = userdata;
+    finish_offscreen(ctx);
+    SDL_Rect *r = &ctx->culling_area;
+    unsigned char *buffer = omf_malloc(r->w * r->h);
+    glReadPixels(r->x, r->y, r->w, r->h, GL_RED, GL_UNSIGNED_BYTE, buffer);
+    surface_create_from_data_flip(dst, r->w, r->h, buffer);
+    surface_set_transparency(dst, -1);
+    omf_free(buffer);
+}
+
+static void capture_screen(void *userdata, video_screenshot_signal screenshot_cb) {
+    gl3_context *ctx = userdata;
+    ctx->screenshot_cb = screenshot_cb;
+}
+
+static void signal_scene_change(void *userdata) {
+    gl3_context *ctx = userdata;
+    atlas_reset(ctx->atlas);
+}
+
+static void signal_draw_atlas(void *userdata, bool toggle) {
+    gl3_context *ctx = userdata;
+    ctx->draw_atlas = toggle;
+}
+
+static void renderer_init(renderer *gl3_renderer) {
+    gl3_renderer->ctx = omf_calloc(1, sizeof(gl3_context));
+}
+
+static void renderer_close(renderer *gl3_renderer) {
+    omf_free(gl3_renderer);
```

Comment:
I think you meant `omf_free(gl3_renderer->ctx)`?

Reply:
Yes, I did. Will fix.

Reply:
Perhaps init and close should be called create and destroy instead?

Reply:
Sounds good to me.